translator/trace/oc_to_otlp.go

```diff
@@ -0,0 +1,420 @@
+				resourceSpans.Spans = append(resourceSpans.Spans, otlpSpan)
+			}
+		}
+	}
+
+	return consumerdata.NewOTLPTraceData(resourceSpanList)
+}
+
+func timestampToUnixnano(ts *timestamp.Timestamp) uint64 {
+	return uint64(internal.TimestampToTime(ts).UnixNano())
+}
+
+func ocSpanToOtlp(ocSpan *octrace.Span) *otlptrace.Span {
+	attrs, droppedAttrCount := ocAttrsToOtlp(ocSpan.Attributes)
+	events, droppedEventCount := ocEventsToOtlp(ocSpan.TimeEvents)
+	links, droppedLinkCount := ocLinksToOtlp(ocSpan.Links)
+
+	childSpanCount := int32(0)
+	if ocSpan.ChildSpanCount != nil {
+		childSpanCount = int32(ocSpan.ChildSpanCount.Value)
+	}
+
+	otlpSpan := &otlptrace.Span{
+		TraceId:                ocSpan.TraceId,
+		SpanId:                 ocSpan.SpanId,
+		Tracestate:             ocTraceStateToOtlp(ocSpan.Tracestate),
+		ParentSpanId:           ocSpan.ParentSpanId,
+		Name:                   truncableStringToStr(ocSpan.Name),
+		Kind:                   ocSpanKindToOtlp(ocSpan.Kind, ocSpan.Attributes),
+		StartTimeUnixnano:      timestampToUnixnano(ocSpan.StartTime),
+		EndTimeUnixnano:        timestampToUnixnano(ocSpan.EndTime),
+		Attributes:             attrs,
+		DroppedAttributesCount: droppedAttrCount,
+		Events:                 events,
+		DroppedEventsCount:     droppedEventCount,
+		Links:                  links,
+		DroppedLinksCount:      droppedLinkCount,
+		Status:                 nil,
+		LocalChildSpanCount:    childSpanCount,
+	}
+
+	return otlpSpan
+}
+
+// Convert tracestate to W3C format. See the https://w3c.github.io/trace-context/
+func ocTraceStateToOtlp(ocTracestate *octrace.Span_Tracestate) string {
+	if ocTracestate == nil {
+		return ""
+	}
+	var sb strings.Builder
+	for i, entry := range ocTracestate.Entries {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(strings.Join([]string{entry.Key, entry.Value}, "="))
+	}
+	return sb.String()
+}
+
+func ocAttrsToOtlp(ocAttrs *octrace.Span_Attributes) (otlpAttrs []*otlpcommon.AttributeKeyValue, droppedCount uint32) {
+	if ocAttrs == nil {
+		return nil, 0
+	}
+
+	otlpAttrs = make([]*otlpcommon.AttributeKeyValue, 0, len(ocAttrs.AttributeMap))
+	for key, ocAttr := range ocAttrs.AttributeMap {
+
+		otlpAttr := &otlpcommon.AttributeKeyValue{Key: key}
+
+		switch attribValue := ocAttr.Value.(type) {
+		case *octrace.AttributeValue_StringValue:
+			otlpAttr.StringValue = truncableStringToStr(attribValue.StringValue)
+			otlpAttr.Type = otlpcommon.AttributeKeyValue_STRING
+
+		case *octrace.AttributeValue_IntValue:
+			otlpAttr.IntValue = attribValue.IntValue
+			otlpAttr.Type = otlpcommon.AttributeKeyValue_INT
+
+		case *octrace.AttributeValue_BoolValue:
+			otlpAttr.BoolValue = attribValue.BoolValue
+			otlpAttr.Type = otlpcommon.AttributeKeyValue_BOOL
+
+		case *octrace.AttributeValue_DoubleValue:
+			otlpAttr.DoubleValue = attribValue.DoubleValue
+			otlpAttr.Type = otlpcommon.AttributeKeyValue_DOUBLE
+
+		default:
+			str := "<Unknown OpenCensus Attribute for key \"" + key + "\">"
```

Comment:
Since key is already present in another field, is it worth having the key name in value as well? Could replace it with a constant.

Reply:
This is copied from Jaeger translator. I can remove it, I think you are right.

Reply:
Removed.